pkg/oauthserver/userregistry/identitymapper/provision_test.go

```diff
@@ -52,241 +57,295 @@ func TestProvision(t *testing.T) {
 		ProviderName     string
 		ProviderUserName string
 
-		ExistingIdentity           *userapi.Identity
-		ExistingUser               *userapi.User
+		ExistingObjects            []runtime.Object
 		NewIdentityGetterResponses []interface{}
 
-		ExpectedActions  []test.Action
+		ValidateActions  func(t *testing.T, actions []clienttesting.Action)
 		ExpectedError    bool
 		ExpectedUserName string
 	}{
 		"no identity, create user succeeds": {
 			ProviderName:     "idp",
 			ProviderUserName: "bob",
 
-			ExistingIdentity: nil,
-			ExistingUser:     nil,
 			NewIdentityGetterResponses: []interface{}{
 				makeUser("bobUserUID", "bob", "idp:bob"),
 			},
 
-			ExpectedActions: []test.Action{
-				{Name: "GetIdentity", Object: "idp:bob"},
+			ValidateActions: func(t *testing.T, actions []clienttesting.Action) {
+				if len(actions) != 2 {
+					t.Fatal(spew.Sdump(actions))
+				}
+				if !actions[0].Matches("get", "identities") || actions[0].(clienttesting.GetAction).GetName() != "idp:bob" {
+					t.Error(spew.Sdump(actions))
+				}
 				// ... new identity user getter creates user
-				{Name: "CreateIdentity", Object: makeIdentity("", "idp", "bob", "bobUserUID", "bob")},
+				if !actions[1].Matches("create", "identities") {
+					t.Error(spew.Sdump(actions))
+				}
+				actual := actions[1].(clienttesting.CreateAction).GetObject().(*userapi.Identity)
+				if expected := makeIdentity("", "idp", "bob", "bobUserUID", "bob"); !equality.Semantic.DeepEqual(expected, actual) {
+					t.Error(diff.ObjectDiff(expected, actual))
+				}
 			},
 			ExpectedUserName: "bob",
 		},
 		"no identity, alreadyexists error retries": {
 			ProviderName:     "idp",
 			ProviderUserName: "bob",
 
-			ExistingIdentity: nil,
-			ExistingUser:     nil,
 			NewIdentityGetterResponses: []interface{}{
 				kerrs.NewAlreadyExists(userapi.Resource("User"), "bob"),
 				makeUser("bobUserUID", "bob", "idp:bob"),
 			},
 
-			ExpectedActions: []test.Action{
-				{Name: "GetIdentity", Object: "idp:bob"},
+			ValidateActions: func(t *testing.T, actions []clienttesting.Action) {
+				if len(actions) != 3 {
+					t.Fatal(spew.Sdump(actions))
+				}
+				if !actions[0].Matches("get", "identities") || actions[0].(clienttesting.GetAction).GetName() != "idp:bob" {
+					t.Error(spew.Sdump(actions))
```

Comment:
Shouldn't this be t.Error(spew.Sdump(actions[0]))
(and similar below ?)

Reply:
Or perhaps even add what action failed exactly to the error, maybe via helper like:
```
func checkAction(num int, actions []clienttesting.Action, match string[]) {
    if !actions[num].Matches(**match) {
        t.Errorf("Expected action %d to match %v, found %s", num, match, spew.Sdump(actions[num]))
    }
}
```
then:
```
    checkAction(actions, 0, ["create", "identities"])
```
and to check names:
```
func checkActionAndName(num int, actions []clienttesting.Action, match string[], name) {
    if !actions[num].Matches(**match) ||  actions[num].(clienttesting.GetAction).GetName() != name {
        t.Errorf("Expected action %d to match %v, found %s", num, match, spew.Sdump(actions[num]))
    }
}
```
then:
```
    checkActionAndName(actions, 0, ["get", "identities"], "idp:bob")
```
and so on...